packages/runtime/src/next/components/tests/page-date-prop-controller.test.tsx

```diff
@@ -0,0 +1,123 @@
+import { randomUUID } from 'crypto'
+import {
+  Date as DatePropController,
+  DateDescriptor,
+  Types,
+  createDatePropControllerDataFromString,
+} from '@makeswift/prop-controllers'
+import { Page } from '../page'
+import { act } from 'react-dom/test-utils'
+import { ReactRuntimeProvider } from '../../context/react-runtime'
+import { ReactRuntime } from '../../../react'
+import { forwardRef } from 'react'
+import {
+  createMakeswiftPageSnapshot,
+  createRootComponent,
+} from '../../../utils/tests/element-data-test-test'
+
+describe('Page', () => {
+  test('can render DatePropController v0 data', async () => {
+    // Arrange
+    const dateDefinitionV0: DateDescriptor = {
+      type: Types.Date,
+      options: {},
+    }
+    const dateData = new Date(Date.now()).toISOString()
+    const TestComponentType = 'TestComponent'
+    const testId = 'test-id'
+    const elementData: ElementData = createRootComponent([
+      {
+        key: randomUUID(),
+        type: TestComponentType,
+        props: {
+          date: createDatePropControllerDataFromString(dateData, dateDefinitionV0),
+        },
+      },
+    ])
+    const snapshot = createMakeswiftPageSnapshot(elementData)
+    const runtime = new ReactRuntime()
+
+    runtime.registerComponent(
+      forwardRef<HTMLDivElement, { date?: string }>(({ date }, ref) => {
+        return (
+          <div ref={ref} data-testid={testId}>
+            {date}
+          </div>
+        )
+      }),
+      {
+        type: TestComponentType,
+        label: 'TestComponent',
+        props: {
+          date: DatePropController(),
+        },
+      },
+    )
+
+    await act(async () =>
+      render(
+        <ReactRuntimeProvider runtime={runtime}>
+          <Page snapshot={snapshot} />
+        </ReactRuntimeProvider>,
+      ),
+    )
+
+    expect(screen.getByTestId(testId)).toHaveTextContent(dateData)
+  })
+
+  test('can render DatePropController v1 data', async () => {
+    // Arrange
+    const dateDefinitionV1: DateDescriptor = {
+      type: Types.Date,
+      version: 1,
+      options: {},
+    }
+    const dateData = new Date(Date.now()).toISOString()
+    const TestComponentType = 'TestComponent'
+    const testId = 'test-id'
+    const elementData: ElementData = createRootComponent([
+      {
+        key: randomUUID(),
+        type: TestComponentType,
+        props: {
+          date: createDatePropControllerDataFromString(dateData, dateDefinitionV1),
+        },
+      },
+    ])
+    const snapshot = createMakeswiftPageSnapshot(elementData)
+    const runtime = new ReactRuntime()
+
+    runtime.registerComponent(
+      forwardRef<HTMLDivElement, { date?: string }>(({ date }, ref) => {
+        return (
+          <div ref={ref} data-testid={testId}>
+            {date}
+          </div>
+        )
+      }),
+      {
+        type: TestComponentType,
+        label: 'TestComponent',
+        props: {
+          date: DatePropController(),
+        },
+      },
+    )
```

Comment:
There's a ton of boilerplate here that looks copy pasted from other test cases in this file which makes it hard to follow these tests. Maybe we can extract them into a utility in this file?

Reply:
I couldn't find a nice abstraction to extract them yet. For each prop-controller, we need to register different prop-controller, and different custom components because sometimes we're testing the style, while sometimes we're just testing the value.

If you have an idea of how we can extract this into a utility file please let me know. I don't think it'll be hard to migrate them later, as we only have two test cases for each prop-controller.